### PR TITLE
integrate withdrawal lock

### PR DESCRIPTION
**Pull request by zeroqn**
Generate withdrawal cells
Revert withdrawal cells

### CODE REVIEW

File: crates/block-producer/src/rpc_client.rs [hunk 1]
```diff
@@ -51,7 +51,7 @@ fn to_jsonh256(v: H256) -> JsonH256 {
     h.into()
 }
 
-fn to_result<T: DeserializeOwned>(output: Output) -> anyhow::Result<T> {
+pub fn to_result<T: DeserializeOwned>(output: Output) -> anyhow::Result<T> {
```
Comment (jjyr) [inline crates/block-producer/src/rpc_client.rs:54]: This is a internal function, should not make it public

File: crates/block-producer/src/transaction_skeleton.rs [hunk 2]
```diff
@@ -80,6 +84,13 @@ impl TransactionSkeleton {
     pub fn signature_entries(&self) -> Vec<SignatureEntry> {
         let mut entries: HashMap<[u8; 32], SignatureEntry> = Default::default();
         for (index, input) in self.inputs.iter().enumerate() {
+            // Skip withdrawal lock witness args
+            if let Some(witness_args) = self.witnesses().get(index) {
```
Comment (jjyr) [inline crates/block-producer/src/transaction_skeleton.rs:88]: Why do we have signed witnesses?

File: crates/block-producer/src/withdrawal.rs [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+        CollectedCustodianCells {
+            cells_info: Default::default(),
+            capacity: 0,
+            sudt: Default::default(),
+            fullfilled_sudt_script: Default::default(),
+        }
+    }
+}
+
+async fn query_finalized_custodian_cells(
```
Comment (jjyr) [inline crates/block-producer/src/withdrawal.rs:444]: Put this function into rpc_client.rs

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/block-producer/src/rpc_client.rs (hunk 1 above)
Comment (zeroqn) [reply crates/block-producer/src/rpc_client.rs:54]: How about pub(crate) or RPCClient::to_result ?
I place some cell query code only used by withdrawal in withdrawal.rs, it needs to access these code.

File: crates/block-producer/src/transaction_skeleton.rs (hunk 2 above)
Comment (zeroqn) [reply crates/block-producer/src/transaction_skeleton.rs:88]: These are not signed withnesses, they are ```UnlockWithdrawalWitness``` and use to revert withdrawal cells back to custodian cells,

File: crates/block-producer/src/withdrawal.rs (hunk 3 above)
Comment (zeroqn) [reply crates/block-producer/src/withdrawal.rs]: Done

File: crates/block-producer/src/rpc_client.rs (hunk 1 above)
Comment (zeroqn) [reply crates/block-producer/src/rpc_client.rs]: Done

File: crates/block-producer/src/withdrawal.rs (hunk 3 above)
Comment (zeroqn) [reply crates/block-producer/src/withdrawal.rs]: Done

File: crates/block-producer/src/rpc_client.rs [hunk 4]
```diff
@@ -89,11 +94,46 @@ fn parse_deposit_request(
...
+        }
+    }
+}
+
 #[derive(Clone)]
 pub struct RPCClient {
     pub indexer_client: HttpClient,
     pub ckb_client: HttpClient,
     pub rollup_type_script: ckb_types::packed::Script,
+    pub rollup_config_cell_dep: gw_types::packed::CellDep,
```
Comment (jjyr) [inline crates/block-producer/src/rpc_client.rs:136]: This field is useless

Comment (zeroqn) [reply crates/block-producer/src/rpc_client.rs]: Take ```rollup_config_cell_dep``` from ```BlockProducerConfig``` now. Done

**Review by jjyr — APPROVED** [review]

**Pull request merged**